serviceDayLookout: add parameter to look more days out in future (or past, for arriveBy=true)

sdjacobs opened the pull request:
Some transit agencies, especially in rural areas, run infrequent service. For example, the Kingdom Shopper 2 run by Rural Community Transit (RCT) in Vermont runs the 2nd and 4th Wednesday of every month. In these cases, it would be good to give a result for trip plans even if the time is quite far from the time that the service runs.

This PR adds an extra parameter, serviceDayLookout, which when specified allows OTP to consider more ServiceDays when looking for transit service. If serviceDayLookout=n, OTP will search out n days in the future for service, or n days in the past for arriveBy=true searches. If serviceDayLookout is _not_ specified, behavior is unchanged.

This PR includes a test of this behavior using RCT data.

gmellemstrand commented:
I've looked through the code and tested this, and it looks good. I'm just wondering how you are using this feature from the client. How do you know when to specify extra servicedays? Is it something you do if the main search does not return any results?

I also tested a search between Oslo and Bergen on the Norwegian network, and search time increased about 50% when specifying a serviceDayLookout of 7 days. Searches within Oslo increased by about the same.

sdjacobs commented:
We're just setting the `serviceDayLookout` to be very large globally (31, since some routes are once a month in Vermont). Since it's a small network, the performance effects weren't a concern.

This is making me think though, maybe a more flexible solution would be to make the change in TransitBoardAlight: https://github.com/opentripplanner/OpenTripPlanner/blob/c9e760bc839ab9e214aad06eb1426c9f7c204810/src/main/java/org/opentripplanner/routing/edgetype/TransitBoardAlight.java#L266
Maybe instead of searching through a specific number of days, if you determine that the nearest day with service is too far away, it could calculate what ServiceDay to instantiate by looking at the service IDs for trips in the TripPattern. This still might have poor performance effects though, I'm not sure. (I would think in both implementations performance would depend on how many TripPatterns have intermittent service.)

gmellemstrand commented:
https://github.com/opentripplanner/OpenTripPlanner/blob/c9e760bc839ab9e214aad06eb1426c9f7c204810/src/main/java/org/opentripplanner/routing/edgetype/TransitBoardAlight.java#L250-L261

It seems like you cannot break after finding a trip, because of overlap near midnight. And possibly ring routes could also complicate this?

I think if you could break after the first trip, you wouldn't get so much of a performance hit from searching a long time into the future.

gmellemstrand commented:
I did some tests and I think what you are thinking about calculating what serviceDays to instantiate is interesting. However, it might be not be worth spending too much time on something that is going to be replaced by a new algorithm anyway. I think the feature as-is is useful to us at Entur, and I see no problem with it being merged into master.

t2gran commented:
I am not sure if I am inline with the rest here, but this is my view on issues and PRs:

For the future it is good to create an Issue and a PR, when the PR contain new feature(s). Then the functionality and high level design should be discussed in the issue and code in the PR.

The group of people interested in issues and PRs are slightly different and the lifecycle, too. An issue may be referenced and document a feature, hence provide insight for a long time after it is closed - a PR more or less when the PR is merged into master.

Any thoughts on this?

abyrd commented:
@t2gran that's how I see it too - an issue describes the problem or need, people can discuss it on the issue, then a PR responds to that issue with a proposed implementation. 

We don't need to be completely strict - for a small fix or change, a detailed description on the PR might be acceptable, but generally it's better for the PR to be a response to an issue (ideally with a commit message somewhere in the PR that says fixes #123, auto-closing the issue when the PR is merged).

drewda commented:
@sdjacobs is this PR still relevant to merging all the VTrans MOD work in to mainline OTP? If so, please update and resolve conflicts -- or let others know how we can help get this ready to merge.

sdjacobs commented:
Thanks @drewda! I just resolved conflicts. This PR is relevant to the VTrans MoD project, but it's orthogonal to GTFS-Flex: we needed this change to show customers trips which may be relevant but far in the future (e.g. a twice-monthly "shopper" route.)

At this point, all that's remaining are the approvals. @gmellemstrand could you take another look? Nothing has changed since your last approval but it's been merged with the dev-1.x branch.

abyrd commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -391,6 +391,10 @@
     @QueryParam("geoidElevation")
     private Boolean geoidElevation;
 
+    /* If trip discovery mode (look out a week) should be used */
```
This comment should be revised. "Discovery mode" is not defined, and this parameter seems to set the number of days (not a specific one-week time period). Javadoc here should probably be very similar to RoutingRequest. serviceDayLookout.

abyrd commented on src/main/java/org/opentripplanner/routing/core/RoutingContext.java:
```diff
@@ -384,6 +384,13 @@ private void setServiceDays() {
                         calendarService, agency.getId()));
+                if (opt.serviceDayLookout > 0) {
+                    ServiceDate sd = serviceDate;
+                    for (int i = 0; i < opt.serviceDayLookout; i++) {
```
This is copying the original ServiceDate and then incrementing/decrementing it before adding it. It looks like this will double-add the day immediately after or before the original ServiceDate (since it's already been added on line 385 or 383 above). Rather than having duplicate logic to handle the default "one day on either side" case and the asymmetric "long lookout on one side of the start date" case, it would make more sense to me if the serviceDayLookout parameter just defaulted to 1 and the default (existing) behavior was just treated as a minimum case of service day lookout.

sdjacobs replied:
I'm happy to make then new behavior the default.

Double-adding won't cause correctness problems because of the use of `addIfNotExists`. If there's a performance penalty, the larger issue with double-adding is that we loop over all feeds and all agencies, but any ServiceDay which is created for a timezone which already exists is redundant and won't end up getting added to the list. agencyId is only used to look up the timezone:
https://github.com/opentripplanner/OpenTripPlanner/blob/4820e9cbc59c77ff832a492b9f1f497bf9834feb/src/main/java/org/opentripplanner/routing/core/ServiceDay.java#L42-L47

Shouldn't be too hard to refactor this.

abyrd replied:
Thanks for the clarification. I haven't looked at this code for a while and was scratching my head over the agency loop (which is apparently mostly to deal with multiple time zones, even though it's frequently said elsewhere that OTP assumes one time zone per graph), but now I realize there's a lot of de-duplication going on in the list of service dates. An additional performance concern is that these dates are being added to a list rather than a set, so each check for duplicates is linear-time. But I think this list is usually so short that the linear-time solution is just as fast.

So the existing implementation looks fine, it's just a bit strange to add one extra day, then re-iterate to add some of the same days. It does very simply handle the fact that you want one day before and N days after though. I'm fine with keeping the existing implementation or trying to generalize it so that even the default is a 1-day lookout, whichever you prefer.

abyrd commented on src/main/java/org/opentripplanner/routing/core/RoutingContext.java:
```diff
@@ -384,6 +384,13 @@ private void setServiceDays() {
+                    ServiceDate sd = serviceDate;
+                    for (int i = 0; i < opt.serviceDayLookout; i++) {
+                        sd = opt.arriveBy ? sd.previous() : sd.next();
+                        addIfNotExists(this.serviceDays, new ServiceDay(graph, sd, calendarService, agency.getId()));
```
This adds only services on one side of the specified search date (after or before, but not both). So this handles cases where the service occurs far in the future or past, but not cases where the service day is multiple calendar days long (e.g. ferry running for multiple days as in Norway). This is probably an intentional choice, but these expected use cases (and cases that are not covered) should be explained in the documentation or Javadoc on the new parameters.

sdjacobs replied:
Hm. Would it address the multiple-day issue if we made the serviceDayLookout symmetric? The tradeoff is that iterating through ServiceDays in `TransitBoardAlight` is a hot spot. We could also add "forwardLookout" and "backwardLookout" parameters, but that would increase the complexity a bit. I guess I lean toward adding both parameters with sane defaults (1) and letting power users change them. Maybe this is something that should just be configurable via `router-config.json` and not via the API resource. Thoughts?

abyrd replied:
I don't even think we necessarily need to handle the very long multi-day trips here. It's just something that came up in Norway that also involves increasing the number of service days, so I think we need to state clearly in docs what problem is being solved here. If it's only boarding trips many days in the future (but not boarding trips many days long) that's fine, I just think it will be helpful to future readers to make that clear.

If you do want to add a backwardLookout, I'd discuss with @t2gran or @gmellemstrand to see if this would work in their use case. 

It is debatable whether this should be changed in router-config or per-request in the API. The main problem I see with doing it in the API is that you can grind the server to a halt with this parameter, so it's a kind of DDOS target. The problem I see with doing it in router-config is that some special trips really need to be multi-day, and most others do not. So it could lead to a ton of extra calculation where not necessary.

sdjacobs commented:
Just pushed some changes. I added some documentation. Also, in reference to the issues @abyrd  and I discussed above, I made some specific decisions (though I'm happy to change back if people disagree):
- Refactored to avoid the double-adding of service days
- Removed from API parameters, to avoid callers locking up the server. This also provides us more flexibility if we want to change this moving forward.
- I didn't add the "backward lookout" we discussed. Looks like @gmellemstrand's code only increases the amount of service days for the specific routes its applicable to. That's a more performant way of handling this, but I thought outside the scope of this PR.

gmellemstrand commented:
The way I see it we have these problems to solve:

1. Show infrequent trips in areas with little transit coverage.
2. Multi-day trips consisting of multiple legs (can be up to 3-4 days in Norway, and longer when we start thinking about multiple countries)
3. Get a timetable view of up to a month at a time
4. Single legs that lasts multiple days and can be boarded at any of those days (Hurtigruta takes 7 days to travel all along the Norwegian coast)

This pull request solves problem 1 and 2 at the cost of about 50% increase in search time when testing with the Norwegian graph. Probably with a combination of running parallell searches for each applicable day for 1 and 3, and loading multiple service days for 2. Case 4 would probably best be handled with special logic for those specific routes.

I agree that the scope of this PR should be to solve 1 and 2. Maybe add a parameter called something like "maxServiceDayLookout" in router-config and then keep "serviceDayLookout" as a request parameter?

sdjacobs commented:
@gmellemstrand @abyrd are there other changes I should make to this PR?

gmellemstrand commented:
I have no other requests for changes before this is merged.

abyrd commented:
Merged! I'm not going to cherry-pick this one onto dev-2.x because it is coupled to the transit routing implementation.
